pkg/controller/dns/managedzone.go

```diff
@@ -0,0 +1,188 @@
+	cr, ok := mg.(*v1alpha1.ManagedZone)
+	if !ok {
+		return nil, errors.New(errNotManagedZone)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderReference.Name}, p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errProviderSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	dnsService, err := dns.NewService(ctx, option.WithCredentialsJSON(s.Data[p.Spec.CredentialsSecretRef.Key]))
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+
+	return &managedZoneExternal{Service: dnsService, kube: c.kube, projectID: p.Spec.ProjectID}, nil
+}
+
+type managedZoneExternal struct {
+	kube client.Client
+	*dns.Service
+	projectID string
+}
+
+func (e *managedZoneExternal) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.ManagedZone)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotManagedZone)
+	}
+
+	if meta.GetExternalName(cr) == "" {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	t, err := e.ManagedZones.Get(e.projectID, meta.GetExternalName(cr)).Do()
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGet)
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	dns2.LateInitialize(&cr.Spec.ForProvider, t)
+	if !cmp.Equal(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdate)
+		}
+	}
+
+	cr.SetConditions(runtimev1alpha1.Available())
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: dns2.IsUpToDate(cr.Spec.ForProvider, *t),
+	}, nil
+
+}
+
+func (e *managedZoneExternal) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.ManagedZone)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotManagedZone)
+	}
+	cr.SetConditions(runtimev1alpha1.Creating())
+
+	resp, err := e.ManagedZones.Create(e.projectID, dns2.GenerateManagedZone(e.projectID, cr.Spec.ForProvider)).Do()
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateManagedZone)
+	}
+
+	meta.SetExternalName(cr, fmt.Sprintf("%d", resp.Id))
+	return managed.ExternalCreation{}, errors.Wrap(e.kube.Update(ctx, cr), errKubeUpdate)
+}
+
+func (e *managedZoneExternal) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	return managed.ExternalUpdate{}, nil
+}
+
+func (e *managedZoneExternal) Delete(ctx context.Context, mg resource.Managed) error {
```

Comment:
Do we need adding finalizer explicitly or crossplane already did this for us?

Reply:
Crossplane did this for us
```
  Finalizers:
    finalizer.managedresource.crossplane.io
```

We just follow it's framework to implement our business logic